### PR TITLE
[AskMode] Generate an AnalzyerDriver exception if OperationWalker stack overflows

### LINKED ISSUE #9025: `InsufficientExecutionStackException` while trying to get Opertion from a long arithmetic expression
Can be repro'd in following tests with OperationWalker hooked-up.
```
Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.CodeGenOperatorTests.EmitSequenceOfBinaryExpressions_01
Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.CodeGenOperatorTests.EmitSequenceOfBinaryExpressions_02
Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.CodeGenOperatorTests.EmitSequenceOfBinaryExpressions_04
Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.CodeGenOperatorTests.EmitSequenceOfBinaryExpressions_06
```

There's an related issue created for compiler side #5395.

### PR DESCRIPTION
**Pull request by genlu**
This should fix #9025.

@JohnHamby @srivatsn @mavasani @heejaechang @AlekseyTs 

Add @MattGertz @ManishJayaswal for ask mode approval.

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -1548,7 +1548,19 @@ private static void ComputeDeclarationsInNode(SemanticModel semanticModel, ISymb
...
+                                    // OperationWalker can throw `InsufficientExecutionStackException`
+                                    operationsToAnalyze = GetOperationsToAnalyze(operationBlocksToAnalyze);
+                                }
+                                catch (Exception ex) when (StackGuard.IsInsufficientExecutionStackException(ex))
+                                {
+                                    var diagnostic = AnalyzerExecutor.CreateDriverExceptionDiagnostic(ex);
+                                    var analyzer = this.analyzers[0];
+
+                                    analyzerExecutor.OnAnalyzerException(ex, analyzer, diagnostic);
+                                }
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I agree with John. We should also probably log a non-fatal watson for diagnosis.

Comment (genlu) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: OK, working on it.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 2]
```diff
@@ -1548,7 +1548,19 @@ private static void ComputeDeclarationsInNode(SemanticModel semanticModel, ISymb
...
                             {
                                 var operationBlocksToAnalyze = GetOperationBlocksToAnalyze(executableCodeBlocks, semanticModel, cancellationToken);
-                                var operationsToAnalyze = GetOperationsToAnalyze(operationBlocksToAnalyze);
+                                var operationsToAnalyze = ImmutableArray<IOperation>.Empty;
+                                try
+                                {
+                                    // OperationWalker can throw `InsufficientExecutionStackException`
+                                    operationsToAnalyze = GetOperationsToAnalyze(operationBlocksToAnalyze);
+                                }
+                                catch (Exception ex) when (FatalError.ReportWithoutCrashUnlessCanceled(ex))
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Do we really want to create a Watson for an InsufficientExecutionStackException?

Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Probably we want the Watson only for non-stack overflow cases?

Comment (genlu) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Fixed.

**Comment by genlu** [comment]
@mavasani Am i using non-fatal-watson correctly? I basically just copied code from other usage of it.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by heejaechang** [comment]
looks fine on non-fatal watson usage.

**Comment by genlu** [comment]
retest vsi please

**Comment by genlu** [comment]
Retest this please.

**Comment by genlu** [comment]
@MattGertz ?

**Comment by MattGertz** [comment]
Approved.

**Pull request merged**